src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -354,104 +459,192 @@ cbuf_mark_relinquish_all(struct cbuf_comp_info *cci)
 	int i;
 	struct cbuf_bin *bin;
 	struct cbuf_info *cbi;
+	struct cbuf_maps *m;
 
 	for(i=cci->nbin-1; i>=0; --i) {
 		bin = &cci->cbufs[i];
 		cbi = bin->c;
 		do {
-			CBUF_FLAG_ADD(cbi->owner.m, CBUF_RELINQ);
-			cbi = FIRST_LIST(cbi, next, prev);
+			if (!cbi) break;
+			m = &cbi->owner;
+			do {
+				CBUF_FLAG_ATOMIC_ADD(m->m, CBUF_RELINQ);
+				m  = FIRST_LIST(m, next, prev);
+			} while(m != &cbi->owner);
+			cbi   = FIRST_LIST(cbi, next, prev);
 		} while (cbi != bin->c);
 	}
 }
+
 static inline void
 cbuf_unmark_relinquish_all(struct cbuf_comp_info *cci)
 {
 	int i;
 	struct cbuf_bin *bin;
 	struct cbuf_info *cbi;
+	struct cbuf_maps *m;
 
 	for(i=cci->nbin-1; i>=0; --i) {
 		bin = &cci->cbufs[i];
 		cbi = bin->c;
 		do {
-			CBUF_FLAG_REM(cbi->owner.m, CBUF_RELINQ);
-			cbi = FIRST_LIST(cbi, next, prev);
+			if (!cbi) break;
+			m = &cbi->owner;
+			do {
+				CBUF_FLAG_ATOMIC_REM(m->m, CBUF_RELINQ);
+				m  = FIRST_LIST(m, next, prev);
+			} while(m != &cbi->owner);
+			cbi   = FIRST_LIST(cbi, next, prev);
 		} while (cbi != bin->c);
 	}
 }
 
+static inline void
+cbuf_thread_block(struct cbuf_comp_info *cci, unsigned long request_size)
+{
+	struct blocked_thd bthd;
+
+	bthd.thd_id       = cos_get_thd_id();
+	bthd.request_size = request_size;
+	rdtscll(bthd.blk_start);
+	ADD_LIST(&cci->bthd_list, &bthd, next, prev);
+	cci->num_blocked_thds++;
+	cbuf_mark_relinquish_all(cci);
+	CBUF_RELEASE();
+	sched_block(cos_spd_id(), 0);
+}
+
+/* wake up all blocked threads whose request size smaller than or equal to available size */
+static void
+cbuf_thd_wake_up(struct cbuf_comp_info *cci, unsigned long sz)
+{
+	struct blocked_thd *bthd, *next;
+	unsigned long long cur, tot;
+
+	assert(cci->num_blocked_thds >= 0);
+	/* Cannot wake up thd when in shrink */
+	assert(cci->target_size >= cci->allocated_size);
+
+	if (cci->num_blocked_thds == 0) return;
+	bthd = cci->bthd_list.next;
+	while (bthd != &cci->bthd_list) {
+		next = FIRST_LIST(bthd, next, prev);
+		if (bthd->request_size <= sz) {
+			REM_LIST(bthd, next, prev);
+			cci->num_blocked_thds--;
+			rdtscll(cur);
```

Comment:
It is not meaningful all the time.
And It depends on how policy use those information